In [13]:
import os 
print(os.getcwd())

try:
    os.chdir("../../workshop_02")

except FileNotFoundError:
    print("Directory not found or already exists.")
    
print(os.getcwd())

/home/edwards/Documentos/university/etl/workshop_02
Directory not found or already exists.
/home/edwards/Documentos/university/etl/workshop_02


In [14]:
import pandas as pd
import numpy as np
import sys
import matplotlib.pyplot as plt

In [15]:
spotify_df= pd.read_csv('data/spotify_dataset.csv')
spotify_df.head(10)

Unnamed: 0                track_id                               artists  \
0           0  5SuOikwiRyPMVoIQDJUgSV                           Gen Hoshino   
1           1  4qPNDBW1i3p13qLCt0Ki3A                          Ben Woodward   
2           2  1iJBSr7s7jYXzM8EGcbK5b                Ingrid Michaelson;ZAYN   
3           3  6lfxq3CG4xtTiEg7opyCyx                          Kina Grannis   
4           4  5vjLSffimiIP26QG5WcN2K                      Chord Overstreet   
5           5  01MVOl9KtVTNfFiBU9I7dc                          Tyrone Wells   
6           6  6Vc5wAMmXdKIAM7WUoEb7N  A Great Big World;Christina Aguilera   
7           7  1EzrEOXmMH3G43AXT1y7pA                            Jason Mraz   
8           8  0IktbUcnAGrvD03AWnz3Q8             Jason Mraz;Colbie Caillat   
9           9  7k9GuJYLp2AzqokyEdwEw2                        Ross Copperman   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   
5                               Days I Will Remember   
6                        Is There Anybody Out There?   
7                We Sing. We Dance. We Steal Things.   
8                We Sing. We Dance. We Steal Things.   
9                                             Hunger   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Help Falling In Love          71       201933     False   
4                     Hold On          82       198853     False   
5        Days I Will Remember          58       214240     False   
6               Say Something          74       229400     False   
7                   I'm Yours          80       242946     False   
8                       Lucky          74       189613     False   
9                      Hunger          56       205594     False   

   danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0         0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1         0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2         0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3         0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4         0.618  0.4430  ...    -9.681     1       0.0526        0.4690   
5         0.688  0.4810  ...    -8.807     1       0.1050        0.2890   
6         0.407  0.1470  ...    -8.822     1       0.0355        0.8570   
7         0.703  0.4440  ...    -9.331     1       0.0417        0.5590   
8         0.625  0.4140  ...    -8.700     1       0.0369        0.2940   
9         0.442  0.6320  ...    -6.770     1       0.0295        0.4260   

   instrumentalness  liveness  valence    tempo  time_signature  track_genre  
0          0.000001    0.3580   0.7150   87.917               4     acoustic  
1          0.000006    0.1010   0.2670   77.489               4     acoustic  
2          0.000000    0.1170   0.1200   76.332               4     acoustic  
3          0.000071    0.1320   0.1430  181.740               3     acoustic  
4          0.000000    0.0829   0.1670  119.949               4     acoustic  
5          0.000000    0.1890   0.6660   98.017               4     acoustic  
6          0.000003    0.0913   0.0765  141.284               3     acoustic  
7          0.000000    0.0973   0.7120  150.960               4     acoustic  
8          0.000000    0.1510   0.6690  130.088               4     acoustic  
9          0.004190    0.0735   0.1960   78.899               4     acoustic  

[10 rows x 21 columns]

In [5]:
spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        114000 non-null  int64  
 1   track_id          114000 non-null  object 
 2   artists           113999 non-null  object 
 3   album_name        113999 non-null  object 
 4   track_name        113999 non-null  object 
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  bool   
 8   danceability      114000 non-null  float64
 9   energy            114000 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          114000 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      114000 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          11

In [6]:
spotify_df.isnull().sum()

Unnamed: 0          0
track_id            0
artists             1
album_name          1
track_name          1
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64

In [7]:
for col in spotify_df.columns:
    print(f"{col}: {spotify_df[col].duplicated().sum()}")

Unnamed: 0: 0
track_id: 24259
artists: 82562
album_name: 67410
track_name: 40391
popularity: 113899
duration_ms: 63303
explicit: 113998
danceability: 112826
energy: 111917
key: 113988
loudness: 94520
mode: 113998
speechiness: 112511
acousticness: 108939
instrumentalness: 108654
liveness: 112278
valence: 112210
tempo: 68347
time_signature: 113995
track_genre: 113886


Vemos que hay varios duplicados, pero entiendo el contexto. Sabemos que un artista puede tener varias canciones, un albúm también puede tener muchas canciones, y así con cosas que pueden ser un poco lógicas de detallar.

**LO ÚNICO QUE NO PUEDE SUCEDER** es que el id de la canción se repita, por lo que lo vamos a eliminar

Y además voy a eliminar columnas que sean demasiado complejas de tratar o que son muy técnicas por asi decirlo y me quedaré con las necesarias y eliminaremos los valores nulos que vimos que se encuentran en las columnas:

+ artists             
+ album_name          
+ track_name

También eliminaremos los valores duplicados en **track_id**

In [16]:
df = spotify_df.drop(columns=['Unnamed: 0', 'key', 'mode', 'speechiness', 'liveness', 'time_signature'])

df = df.drop_duplicates(subset=['track_id'])

# Drop rows with null values in specified columns
df = df.dropna(subset=['artists', 'album_name', 'track_name'])

# Save the cleaned dataset
output_path = 'data/processed/spotify_dataset_cleaned.csv'
df.to_csv(output_path, index=False)

print(f"ruta al archivo: {output_path}")

ruta al archivo: data/processed/spotify_dataset_cleaned.csv


A partir de aquí trabajaremos con el csv limpio

In [9]:
df = pd.read_csv('data/processed/spotify_dataset_cleaned.csv')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89740 entries, 0 to 89739
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   track_id          89740 non-null  object 
 1   artists           89740 non-null  object 
 2   album_name        89740 non-null  object 
 3   track_name        89740 non-null  object 
 4   popularity        89740 non-null  int64  
 5   duration_ms       89740 non-null  int64  
 6   explicit          89740 non-null  bool   
 7   danceability      89740 non-null  float64
 8   energy            89740 non-null  float64
 9   loudness          89740 non-null  float64
 10  acousticness      89740 non-null  float64
 11  instrumentalness  89740 non-null  float64
 12  valence           89740 non-null  float64
 13  tempo             89740 non-null  float64
 14  track_genre       89740 non-null  object 
dtypes: bool(1), float64(7), int64(2), object(5)
memory usage: 9.7+ MB


Luego de la eliminacion de canciones nos quedaron 89740 canciones

In [11]:
pd.set_option('display.float_format', '{:.2f}'.format)
df.describe()

popularity  duration_ms  danceability   energy  loudness  acousticness  \
count    89740.00     89740.00      89740.00 89740.00  89740.00      89740.00   
mean        33.20    229144.37          0.56     0.63     -8.50          0.33   
std         20.58    112945.78          0.18     0.26      5.22          0.34   
min          0.00      8586.00          0.00     0.00    -49.53          0.00   
25%         19.00    173040.00          0.45     0.46    -10.32          0.02   
50%         33.00    213295.50          0.58     0.68     -7.18          0.19   
75%         49.00    264293.00          0.69     0.85     -5.11          0.62   
max        100.00   5237295.00          0.98     1.00      4.53          1.00   

       instrumentalness  valence    tempo  
count          89740.00 89740.00 89740.00  
mean               0.17     0.47   122.06  
std                0.32     0.26    30.12  
min                0.00     0.00     0.00  
25%                0.00     0.25    99.26  
50%                0.00     0.46   122.01  
75%                0.10     0.68   140.08  
max                1.00     0.99   243.37

In [12]:
df.isnull().sum()

track_id            0
artists             0
album_name          0
track_name          0
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
loudness            0
acousticness        0
instrumentalness    0
valence             0
tempo               0
track_genre         0
dtype: int64

Ya no cuenta con valores nulos, y los duplicados ya hablamos de ellos más atrás

In [13]:
df.head()

track_id                 artists  \
0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Help Falling In Love          71       201933     False   
4                     Hold On          82       198853     False   

   danceability  energy  loudness  acousticness  instrumentalness  valence  \
0          0.68    0.46     -6.75          0.03              0.00     0.71   
1          0.42    0.17    -17.23          0.92              0.00     0.27   
2          0.44    0.36     -9.73          0.21              0.00     0.12   
3          0.27    0.06    -18.52          0.91              0.00     0.14   
4          0.62    0.44     -9.68          0.47              0.00     0.17   

   tempo track_genre  
0  87.92    acoustic  
1  77.49    acoustic  
2  76.33    acoustic  
3 181.74    acoustic  
4 119.95    acoustic

In [14]:
df.nunique()

track_id            89740
artists             31437
album_name          46589
track_name          73608
popularity            101
duration_ms         50696
explicit                2
danceability         1174
energy               2083
loudness            19480
acousticness         5061
instrumentalness     5346
valence              1790
tempo               45652
track_genre           113
dtype: int64

Vemos en generos hay 113 valores únicos, por lo que existen 113 generos, son demasiados para hacer un análisis asi que los vamos a reducir a unos 20

In [15]:
df["track_genre"].unique()

array(['acoustic', 'afrobeat', 'alt-rock', 'alternative', 'ambient',
       'anime', 'black-metal', 'bluegrass', 'blues', 'brazil',
       'breakbeat', 'british', 'cantopop', 'chicago-house', 'children',
       'chill', 'classical', 'club', 'comedy', 'country', 'dance',
       'dancehall', 'death-metal', 'deep-house', 'detroit-techno',
       'disco', 'disney', 'drum-and-bass', 'dub', 'dubstep', 'edm',
       'electro', 'electronic', 'emo', 'folk', 'forro', 'french', 'funk',
       'garage', 'german', 'gospel', 'goth', 'grindcore', 'groove',
       'grunge', 'guitar', 'happy', 'hard-rock', 'hardcore', 'hardstyle',
       'heavy-metal', 'hip-hop', 'honky-tonk', 'house', 'idm', 'indian',
       'indie-pop', 'indie', 'industrial', 'iranian', 'j-dance', 'j-idol',
       'j-pop', 'j-rock', 'jazz', 'k-pop', 'kids', 'latin', 'latino',
       'malay', 'mandopop', 'metal', 'metalcore', 'minimal-techno', 'mpb',
       'new-age', 'opera', 'pagode', 'party', 'piano', 'pop-film', 'pop',
       'pow

Esos son todos los generos que existen, los vamos a encerrar en categorías

In [17]:
genre_categories = {
    'Rock': ['alt-rock', 'alternative', 'emo', 'goth', 'grunge', 'hard-rock', 'indie', 'punk-rock', 'punk', 'psych-rock', 'rock', 'rock-n-roll', 'rockabilly'],
    'Pop': ['cantopop', 'indie-pop', 'j-pop', 'k-pop', 'mandopop', 'pop', 'pop-film', 'power-pop', 'synth-pop'],
    'Electronic': ['breakbeat', 'chicago-house', 'club', 'dance', 'deep-house', 'detroit-techno', 'disco', 'drum-and-bass', 'dub', 'dubstep', 'edm', 'electro', 'electronic', 'garage', 'hardstyle', 'house', 'idm', 'minimal-techno', 'progressive-house', 'techno', 'trance', 'trip-hop'],
    'Hip-Hop/R&B': ['hip-hop', 'r-n-b', 'soul'],
    'Metal': ['black-metal', 'death-metal', 'grindcore', 'heavy-metal', 'metal', 'metalcore'],
    'Folk/Acoustic': ['acoustic', 'bluegrass', 'folk', 'singer-songwriter'],
    'Jazz/Blues': ['blues', 'jazz'],
    'Classical/Instrumental': ['classical', 'new-age', 'opera', 'piano', 'sleep', 'study'],
    'Latin': ['forro', 'latin', 'latino', 'mpb', 'pagode', 'salsa', 'samba', 'sertanejo', 'tango'],
    'Reggae/Dancehall': ['dancehall', 'reggae', 'reggaeton'],
    'Country': ['country', 'honky-tonk'],
    'World/Regional': ['afrobeat', 'brazil', 'french', 'german', 'indian', 'iranian', 'malay', 'spanish', 'swedish', 'turkish', 'world-music'],
    'Anime/Japanese': ['anime', 'j-dance', 'j-idol', 'j-rock'],
    'Kids/Comedy': ['children', 'comedy', 'disney', 'kids', 'show-tunes'],
    'Other': ['ambient', 'british', 'chill', 'funk', 'gospel', 'groove', 'guitar', 'happy', 'hardcore', 'party', 'romance', 'sad', 'ska']
}

genre_to_category = {genre: category for category, genres in genre_categories.items() for genre in genres}

df['genre_category'] = df['track_genre'].map(genre_to_category).fillna('Other')

df = df.drop(columns=['track_genre'])

print(df['genre_category'].unique())

['Folk/Acoustic' 'World/Regional' 'Rock' 'Other' 'Anime/Japanese' 'Metal'
 'Jazz/Blues' 'Electronic' 'Pop' 'Kids/Comedy' 'Classical/Instrumental'
 'Country' 'Reggae/Dancehall' 'Latin' 'Hip-Hop/R&B']


In [7]:
df["genre_category"].nunique()

15

En total encerramos los géneros en 15 categorías

In [18]:
object_cols = df.select_dtypes(include=['object']).columns
for col in object_cols:
    df[col] = df[col].str.lower().str.strip()

Normalizamos las columnas de tipo object para evitar errores a la hora de cualquier análisis, lo haremos en cada uno de los datasets. Solo faltarían los otros 2

+ grammys
+ api

In [19]:
df['duration_min'] = df['duration_ms'] / 60000.
df['duration_min'] = df['duration_min'].round(2)

df = df.drop(columns=['duration_ms'])

Cambiamos la columna a duration_ms por duration_min para una mejor comprensión

In [21]:
df = df.drop(columns=['loudness', 'acousticness', 'instrumentalness', 'valence', 'tempo'])

Eliminamos columnas complejas de tratar, ya que el análisis que buscamos no va en esta dirección. Realmente nos interesa son los artistas, popularidad, followers y demás, y si las canciones con más energía y bailables tienden hacer más populares, al igual que las explicitas o no explicitas

In [23]:
df.head()

track_id                 artists  \
0  5suoikwirypmvoiqdjugsv             gen hoshino   
1  4qpndbw1i3p13qlct0ki3a            ben woodward   
2  1ijbsr7s7jyxzm8egcbk5b  ingrid michaelson;zayn   
3  6lfxq3cg4xttieg7opycyx            kina grannis   
4  5vjlsffimiip26qg5wcn2k        chord overstreet   

                                          album_name  \
0                                             comedy   
1                                   ghost (acoustic)   
2                                     to begin again   
3  crazy rich asians (original motion picture sou...   
4                                            hold on   

                   track_name  popularity  explicit  danceability  energy  \
0                      comedy          73     False         0.676  0.4610   
1            ghost - acoustic          55     False         0.420  0.1660   
2              to begin again          57     False         0.438  0.3590   
3  can't help falling in love          71     False         0.266  0.0596   
4                     hold on          82     False         0.618  0.4430   

  genre_category  duration_min  
0  folk/acoustic          3.84  
1  folk/acoustic          2.49  
2  folk/acoustic          3.51  
3  folk/acoustic          3.37  
4  folk/acoustic          3.31

In [24]:
df.to_csv('data/processed/spotify_dataset_cleaned.csv', index=False)